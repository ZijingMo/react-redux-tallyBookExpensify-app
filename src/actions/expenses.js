import uuid from 'uuid'

// These are the 'action' objects down below

// ADD_EXPENSE
export const addExpense = (
  { 
  description = '', 
  note = '', 
  amount = 0, 
  createAt = 0 
  } = {}
) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createAt
  }
});

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
type: 'REMOVE_EXPENSE',
id
});

// EDIT_EXPENSE
// For the arguments, 'id' is a string number an 'updates' is an object
export const editExpense = (id, updates) => ({
type: 'EDIT_EXPENSE',
id,
updates
});