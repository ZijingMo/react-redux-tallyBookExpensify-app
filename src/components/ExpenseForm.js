import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


//const now = moment();
//console.log(now.format("MMM Do, YYYY"));


// This state component creates an input form, date picker ,and a button for expense data record
export default class ExpenseForm extends React.Component {
  // Using the constructor function to access the expense data which has passed down.If not, the form just shows the default value
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createAt: props.expense ? moment(props.expense.createAt) : moment(),
      calenderFocused: false,
      error: ''
    };
  }
  // These two functions below control the change of input value. It binds the input value and event handlers 'onChange'.
  // If user input some value, property 'description' or 'note' in object 'state' will be updated
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  // This function applies the reg-expression to stipulate the format of amount input, which is any number.two-digit number in this case
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createAt) => {
    if (createAt) {
      this.setState(() => ({ createAt }));
    }
  };
  // Destructuring the object from singleDatePicker component
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  }; 
  // Submitting data to Redux state container and error control
  onSubmit = (e) => {
    // Prevent refreshing the page when user clicks the 'Add Expense' button
    e.preventDefault(); 
    if (!this.state.description || !this.state.amount) {
      // Set error state 
      this.setState(() => ({ error: 'Please provide descriptions and amount.' }));
    } else {
      // Clear the error
      this.setState(() => ({ error: '' }));
      // Add the form (user's input data) to Redux store as an object
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createAt: this.state.createAt.valueOf(),// Using 'moment' library's build-in function valueOf() to convert the time in Unix Timestamp mode (milliseconds)
        note: this.state.note
      });
    }
  };
  render () {
    return (
      <div> 
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input 
            type = "text"
            placeholder ="Description"
            autoFocus
            value = {this.state.description}
            onChange = {this.onDescriptionChange}
          />
          <input 
            type = "text"
            placeholder = "Amount"
            value = {this.state.amount}
            onChange = {this.onAmountChange}
          />
          {/* Third party libraries' component to implement date-pick and calender select */}
          <SingleDatePicker 
            date={this.state.createAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            isDayHighlighted={() => true}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value = {this.state.note}
            onChange = {this.onNoteChange}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}