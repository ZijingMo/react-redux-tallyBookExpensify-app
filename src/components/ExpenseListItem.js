import React from 'react';
import { Link } from 'react-router-dom';

// Destructuring the object to 'dispatch', 'id', 'description', 'amount', 'createAt'
// The component is injected to the 'dispatch' function when we use connect() hoc to wrap it up
const ExpenseListItem = ({ id, description, amount, createAt }) => (
  <div>
    {/* Import 'Link' component to redirect edit page */}
    <Link to={`/edit/${id}`}>
       <h2> { description } </h2> 
    </Link>
    <p> {amount} - {createAt} </p>
  </div>
);

export default ExpenseListItem; 