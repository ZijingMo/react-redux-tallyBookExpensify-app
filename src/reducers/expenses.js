// Expenses Reducer
const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id); // If the accessed id match state's id, that id is bye-bye
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          }; //Spread syntax updates the state (keep unchanged value and change the fluctuating value)
        } else {
          return expense
        };
      }); 
    default: 
      return state;
  }
};

export default expensesReducer;