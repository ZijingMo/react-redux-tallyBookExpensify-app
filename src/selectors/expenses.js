import moment from 'moment';

// Get visible expenses
// First parameter is only object 'expenses', no destruction
// Second parameter destructs the property from object 'filter'
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const createAtMoment = moment(expense.createAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createAt < b.createAt ? 1 : -1 // It means the most recent created date shows on the top
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1 // It means the most expensive amount shows on the top
    }
  });
}

export default getVisibleExpenses;