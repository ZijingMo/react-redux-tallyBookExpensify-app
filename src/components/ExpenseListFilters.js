import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

// This component implements value-input normally. It is able to update change of the input and fire the 'filter''s actions with the help of onChange() method
// The syntaxes of component 'DateRangePicker' follow the 'airbnb/react-dates' documentation
export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render () {
    return (
      <div>
        <input 
          type="text" 
          value={this.props.filters.text} 
          onChange={this.onTextChange} 
        />
    
        <select
          value={this.props.filters.sortBy} 
          onChange={this.onSortChange}
        >
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>

        <DateRangePicker 
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
    filters: state.filters 
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);