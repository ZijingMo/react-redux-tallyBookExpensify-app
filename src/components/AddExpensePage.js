import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// Refactoring the code, from stateless to state component for jest test
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    // This method leads to redirect to the expensify dashboard (Home-Page) when the user click the button 'Add Expense' 
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container"> 
          <ExpenseForm 
            onSubmit = {this.onSubmit}
          />        
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);