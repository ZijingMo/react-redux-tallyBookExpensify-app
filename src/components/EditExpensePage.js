import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

// Refactoring the code, from functional to class component:
// For jest test

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    // This method leads to redirect to the expensify dashboard (Home-Page) when the user click the button 'Add Expense' 
    this.props.history.push('/');
    //console.log('updated', expense);
  };

  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense = {this.props.expense} 
          onSubmit = {this.onSubmit}
        />
        <button 
          onClick = {this.onRemove}
        > 
            Remove 
        </button>
    </div>
    );
  }
}

// New component will subscribe to Redux store updates. 
// This means that any time the store is updated, mapStateToProps will be called. 
// The results of mapStateToProps must be a plain object, which will be merged into the component’s props.
const mapStateToProps = (state, props) => {
  return {
    // It is going to search the array 'expenses' (a lot of objects 'expense' form array 'expenses') in Redux store to match the updated expense the user input
    // 'expense.id' means the id of array 'expenses' in Redux store
    // 'prop.match.params.id' means the property 'id' of an updated object 
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

// Every action creator wrapped into a dispatch call so they may be invoked directly, will be merged into the component’s props.
// Import actions (editExpense, removeExpense) from action folder and subscribe to prop in class 'EditExpensePage'
const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data)) 
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);



// Functional form as a reference

// const EditExpensePage = (props) => {
//   //console.log(props);
//   return (
//     <div>
//       <ExpenseForm
//         expense={props.expense} 
//         onSubmit={(expense) => {
//           // Dispatch the action to edit the expense and redirect to the dashboard
//           props.dispatch(editExpense(props.expense.id, expense));
//           props.history.push('/');
//           console.log('updated', expense);
//         }}
//       />
//       <button 
//         onClick={() => {
//           props.dispatch(removeExpense({id: props.expense.id})); //The argument of action function removeExpense() is an object
//           props.history.push('/');
//         }}
//       > 
//           Remove 
//       </button>
//     </div>
//   );
// };

