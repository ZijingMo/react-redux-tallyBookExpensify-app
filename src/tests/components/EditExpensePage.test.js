import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

// Setting some global variables in this file 
// Applying 'beforeEach()' method to define mock function, root node render (shallow) 
let editExpense, history, wrapper, removeExpense;
beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
  <EditExpensePage 
    editExpense={editExpense} 
    removeExpense ={removeExpense}
    history={history}
    expense={expenses[1]}
  />);
}); 

test('should render EditExpensePage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle action editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('should handle action removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({
    id: expenses[1].id
  });
});

