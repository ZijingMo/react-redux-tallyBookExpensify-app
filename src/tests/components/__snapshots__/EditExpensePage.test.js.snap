// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 2612,
        "createAt": -10544400000,
        "description": "Gas",
        "id": "opi945",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
